Fix Spell Stacking

OfficialExouls opened the pull request:
I believe this should stop people from being able to cast multiple spells as well as account for the additive nature of the modify-stat command. what I mean by that is that the command, instead of setting the stat to the value you give it, it adds the value to the current stat, creating some unintended levels of speed. review it and deem if it's necessary to merge into the base repository.

OfficialExouls committed:
Fix Spell Stacking

edited to account for the additive nature of the modify-stat command

OfficialExouls committed:
Fix Spell Stacking

edited to account for the additive nature of the modify-stat command

OfficialExouls committed:
Fix Spell Stacking

edited to account for the additive nature of the modify-stat command

mdingena commented on src/voodoo/spellbook/spells/haste.ts:
```diff
@@ -27,7 +27,7 @@ export const haste: SpellFunction = async (voodoo, accountId, upgradeConfigs) =>
   });
 
-  const multiplier = 1 + attributes.intensify / 100;
+  const multiplier = attributes.intensify / 100;
```
Shouldn't call this a multiplier anymore if you remove the `1`.

```suggestion
  const bonus = attributes.intensify / 100;
```

mdingena commented on src/voodoo/spellbook/spells/haste.ts:
```diff
@@ -42,8 +42,9 @@ export const haste: SpellFunction = async (voodoo, accountId, upgradeConfigs) =>
+    const currentSpeed = await voodoo.getPlayerCheckStatCurrent({accountId: playerId, stat: 'speed' });
 
-    if (baseSpeed) {
+    if (baseSpeed == currentSpeed) {
```
I'd say that if a user is already under the influence of a normal speed potion but has a Haste invocation that would make them go even faster, Voodoo would make up for the difference.

With this current proposed change, once under the influence of any other speed effect, Haste does not work _at all_.

I'd suggest this

```suggestion
    const buffedSpeed = baseSpeed * (1 + bonus);
    const speedDelta = buffedSpeed - currentSpeed;

    if (speedDelta > 0) {
```

mdingena commented on src/voodoo/spellbook/spells/haste.ts:
```diff
@@ -42,8 +42,9 @@ export const haste: SpellFunction = async (voodoo, accountId, upgradeConfigs) =>
 
-    if (baseSpeed) {
+    if (baseSpeed == currentSpeed) {
       const buffedSpeed = baseSpeed * multiplier;
```
Shouldn't call this the "buffed speed" anymore if it only represents an addition, not the new absolute speed stat.

```suggestion
      const speedBuff = baseSpeed * bonus;
```

There is more code below that is invalidated by these suggestions that you'd have to change for this PR.

mdingena commented on src/voodoo/spellbook/spells/haste.ts:
```diff
@@ -42,8 +42,9 @@ export const haste: SpellFunction = async (voodoo, accountId, upgradeConfigs) =>
 
   for (const playerId of playerIds) {
     const baseSpeed = await voodoo.getPlayerCheckStatBase({ accountId: playerId, stat: 'speed' });
+    const currentSpeed = await voodoo.getPlayerCheckStatCurrent({accountId: playerId, stat: 'speed' });
```
Because this represents two consecutive calls to `player check-stat`, I'd introduce a new method that gets both of these in a single call.

OfficialExouls commented on src/voodoo/spellbook/spells/stoneskin.ts:
```diff
@@ -35,7 +35,7 @@ export const stoneskin: SpellFunction = async (voodoo, accountId, upgradeConfigs
   });
 
-  const multiplier = 1 + attributes.intensify / 100;
+  const multiplier = attributes.intensify / 100;
```
Implementing mdingena's suggestions
```suggestion
  const bonus = attributes.intensify / 100;
```

OfficialExouls commented on src/voodoo/spellbook/spells/stoneskin.ts:
```diff
@@ -50,8 +50,9 @@ export const stoneskin: SpellFunction = async (voodoo, accountId, upgradeConfigs
+    const currentDamageProtection = await voodoo.getPlayerCheckStatCurrent({ accountId: playerId, stat: 'damageprotection' })
 
-    if (baseDamageProtection) {
+    if (baseDamageProtection == currentDamageProtection) {
```
Implementing mdigena's suggestions
```suggestion
    const buffedDamageProtection = baseSpeed * (1 + bonus);
    
    const damageprotectionDelta = buffedDamageProtection - currentDamageProtection;
    
    
    if (damageprotectionDelta > 0) {
```

OfficialExouls commented on src/voodoo/spellbook/spells/stoneskin.ts:
```diff
@@ -50,8 +50,9 @@ export const stoneskin: SpellFunction = async (voodoo, accountId, upgradeConfigs
 
-    if (baseDamageProtection) {
+    if (baseDamageProtection == currentDamageProtection) {
       const buffedDamageProtection = baseDamageProtection * multiplier;
```
Implementing mdigena's suggestions
```suggestion
      const damageprotectionBuff = baseDamageProtection * bonus;
```

OfficialExouls commented on src/voodoo/spellbook/spells/trueStrike.ts:
```diff
@@ -27,7 +27,7 @@ export const trueStrike: SpellFunction = async (voodoo, accountId, upgradeConfig
   });
 
-  const multiplier = 1 + attributes.intensify / 100;
+  const multiplier = attributes.intensify / 100;
```
Implementing mdigena's suggestions
```suggestion
  const bonus = attributes.intensify / 100;
```

OfficialExouls commented on src/voodoo/spellbook/spells/trueStrike.ts:
```diff
@@ -42,6 +42,7 @@ export const trueStrike: SpellFunction = async (voodoo, accountId, upgradeConfig
     const baseDamage = await voodoo.getPlayerCheckStatBase({ accountId: playerId, stat: 'damage' });
+    const currentDamage = await voodoo.getPlayerCheckStatCurrent({ accountId: playerId, stat: 'damage' })
 
     if (baseDamage) {
```
Implementing mdigena's suggestions
```suggestion
    const buffedDamage = baseDamage * (1 + bonus);
    
    const damageDelta = buffedDamage - currentDamage;
    
    
    if (damageDelta > 0) {
```

OfficialExouls commented on src/voodoo/spellbook/spells/trueStrike.ts:
```diff
@@ -42,6 +42,7 @@ export const trueStrike: SpellFunction = async (voodoo, accountId, upgradeConfig
+    const currentDamage = await voodoo.getPlayerCheckStatCurrent({ accountId: playerId, stat: 'damage' })
 
     if (baseDamage) {
       const buffedDamage = baseDamage * multiplier;
```
Implementing mdigena's suggestions
```suggestion
      const damageBuff = baseDamage * bonus;
```

OfficialExouls committed:
Cleaning up and implementing mdigena's suggestions

Co-authored-by: Marc Dingena <marc.dingena@gmail.com>

OfficialExouls committed:
Added general getPlayerCheckStat Method

A general getPlayerCheckStat Method that returns both base and current values

OfficialExouls committed:
update adjuration spells

update haste, stoneskin, and truestrike to implement new getPlayerCheckStat method and variables within the files

OfficialExouls commented on src/voodoo/spellbook/spells/haste.ts:
```diff
@@ -41,14 +41,18 @@ export const haste: SpellFunction = async (voodoo, accountId, upgradeConfigs) =>
 
-    if (baseSpeed) {
-      const buffedSpeed = baseSpeed * multiplier;
+    const buffedSpeed = playerSpeed.base * (1 + bonus);
```
```suggestion
    const buffedSpeed = playerSpeed.Base * (1 + bonus);
```

OfficialExouls commented on src/voodoo/spellbook/spells/haste.ts:
```diff
@@ -41,14 +41,18 @@ export const haste: SpellFunction = async (voodoo, accountId, upgradeConfigs) =>
-      const buffedSpeed = baseSpeed * multiplier;
+    const buffedSpeed = playerSpeed.base * (1 + bonus);
+    
+    const speedDelta = buffedSpeed - playerSpeed.value;
```
```suggestion
    const speedDelta = buffedSpeed - playerSpeed.Value;
```

OfficialExouls commented on src/voodoo/spellbook/spells/haste.ts:
```diff
@@ -41,14 +41,18 @@ export const haste: SpellFunction = async (voodoo, accountId, upgradeConfigs) =>
+    const speedDelta = buffedSpeed - playerSpeed.value;
+
+    if (speedDelta > 0) {
+      const speedBuff = playerSpeed.base * bonus;
```
```suggestion
      const speedBuff = playerSpeed.Base * bonus;
```

OfficialExouls commented on src/voodoo/spellbook/spells/stoneskin.ts:
```diff
@@ -49,14 +49,19 @@ export const stoneskin: SpellFunction = async (voodoo, accountId, upgradeConfigs
 
-    if (baseDamageProtection) {
-      const buffedDamageProtection = baseDamageProtection * multiplier;
+    const buffedDamageProtection = playerDamageProtection.base * (1 + bonus);
```
```suggestion
    const buffedDamageProtection = playerDamageProtection.Base * (1 + bonus);
```

OfficialExouls commented on src/voodoo/spellbook/spells/stoneskin.ts:
```diff
@@ -49,14 +49,19 @@ export const stoneskin: SpellFunction = async (voodoo, accountId, upgradeConfigs
-      const buffedDamageProtection = baseDamageProtection * multiplier;
+    const buffedDamageProtection = playerDamageProtection.base * (1 + bonus);
+    
+    const damageprotectionDelta = buffedDamageProtection - playerDamageProtection.value;
```
```suggestion
    const damageprotectionDelta = buffedDamageProtection - playerDamageProtection.Value;
```

OfficialExouls commented on src/voodoo/spellbook/spells/stoneskin.ts:
```diff
@@ -49,14 +49,19 @@ export const stoneskin: SpellFunction = async (voodoo, accountId, upgradeConfigs
+    
+    
+    if (damageprotectionDelta > 0) {
+      const damageprotectionBuff = playerDamageProtection.base * bonus;
```
```suggestion
      const damageprotectionBuff = playerDamageProtection.Base * bonus;
```

OfficialExouls commented on src/voodoo/spellbook/spells/trueStrike.ts:
```diff
@@ -41,14 +41,19 @@ export const trueStrike: SpellFunction = async (voodoo, accountId, upgradeConfig
 
-    if (baseDamage) {
-      const buffedDamage = baseDamage * multiplier;
+    const buffedDamage = playerDamage.base * (1 + bonus);
```
```suggestion
    const buffedDamage = playerDamage.Base * (1 + bonus);
```

OfficialExouls commented on src/voodoo/spellbook/spells/trueStrike.ts:
```diff
@@ -41,14 +41,19 @@ export const trueStrike: SpellFunction = async (voodoo, accountId, upgradeConfig
-      const buffedDamage = baseDamage * multiplier;
+    const buffedDamage = playerDamage.base * (1 + bonus);
+    
+    const damageDelta = buffedDamage - playerDamage.value;
```
```suggestion
    const damageDelta = buffedDamage - playerDamage.Value;
```

OfficialExouls commented on src/voodoo/spellbook/spells/trueStrike.ts:
```diff
@@ -41,14 +41,19 @@ export const trueStrike: SpellFunction = async (voodoo, accountId, upgradeConfig
+    
+    
+    if (damageDelta > 0) {
+      const damageBuff = playerDamage.base * bonus;
```
```suggestion
      const damageBuff = playerDamage.Base * bonus;
```

OfficialExouls committed:
Fix Variable Names

mdingena committed:
Contibuting to PR #17 (#2)

* Add types

* Refactor abjuration code

* Check correct player stat

Approved PR for PR-17 changing the style of the code to more closely match the dev's.

mdingena reviewed (approved):
Nice work, mate! Thanks for the PR.
